firedrake/cython/dmswarm.pyx

```diff
@@ -0,0 +1,504 @@
+# cython: language_level=3
+
+# Utility functions to derive global and local numbering from DMSwarm
+import cython
+import numpy as np
+from firedrake.petsc import PETSc
+from mpi4py import MPI
+from pyop2.datatypes import IntType
+from libc.string cimport memset
+from libc.stdlib cimport qsort
+cimport numpy as np
+cimport mpi4py.MPI as MPI
+cimport petsc4py.PETSc as PETSc
+
+np.import_array()
+
+include "petschdr.pxi"
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def mark_entity_classes(PETSc.DM swarm):
+    """Mark all points in a given DMSwarm according to the PyOP2 entity
+    classes:
+
+    core   : owned and not in send halo
+    owned  : owned and in send halo
+    ghost  : in halo
+
+    ..note:: this is a copy and paste job from the equivalent function
+             in dmplex.pyx with the relevant modifications made for a
+             DMSwarm.
+
+    :arg swarm: The DMSwarm object encapsulating the mesh topology
+    """
+    cdef:
+        PetscInt pStart, pEnd, cStart, cEnd
+        PetscInt c, ci, p
+        PetscInt nleaves
+        const PetscInt *ilocal = NULL
+        PetscBool non_exec
+        const PetscSFNode *iremote = NULL
+        PETSc.SF point_sf = None
+        PetscBool is_ghost, is_owned
+        DMLabel lbl_core, lbl_owned, lbl_ghost
+
+    pStart = 0
+    pEnd = swarm.getLocalSize()
+
+    swarm.createLabel("pyop2_core")
+    swarm.createLabel("pyop2_owned")
+    swarm.createLabel("pyop2_ghost")
+
+    CHKERR(DMGetLabel(swarm.dm, b"pyop2_core", &lbl_core))
+    CHKERR(DMGetLabel(swarm.dm, b"pyop2_owned", &lbl_owned))
+    CHKERR(DMGetLabel(swarm.dm, b"pyop2_ghost", &lbl_ghost))
+
+    if swarm.comm.size > 1:
+        # Mark ghosts from point overlap SF
+        point_sf = swarm.getPointSF()
+        CHKERR(PetscSFGetGraph(point_sf.sf, NULL, &nleaves, &ilocal, NULL))
+        for p in range(nleaves):
+            CHKERR(DMLabelSetValue(lbl_ghost, ilocal[p], 1))
+    else:
+        # If sequential mark all points as core
+        for p in range(pStart, pEnd):
+            CHKERR(DMLabelSetValue(lbl_core, p, 1))
+        return
+
+    CHKERR(DMLabelCreateIndex(lbl_ghost, pStart, pEnd))
+
+    # Mark all remaining points as core
```

Comment:
This copy-pasted from what is done in `dmplex.pyx`. If it's wrong here it's wrong there too.

Reply:
The relevant code from dmplex.pyx does:

```
    for c in range(cStart, cEnd):
        CHKERR(DMPlexGetTransitiveClosure(plex.dm, c,
                                          PETSC_TRUE,
                                          &nclosure,
                                          &closure))
        is_owned = PETSC_FALSE
        for ci in range(nclosure):
            p = closure[2*ci]
            CHKERR(DMLabelHasPoint(lbl_ghost, p, &is_ghost))
            if is_ghost:
                is_owned = PETSC_TRUE
                break
        if is_owned:
            for ci in range(nclosure):
                p = closure[2*ci]
                CHKERR(DMLabelHasPoint(lbl_ghost, p, &is_ghost))
                if not is_ghost:
                    CHKERR(DMLabelSetValue(lbl_owned, p, 1))
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ 

    if closure != NULL:
        CHKERR(DMPlexRestoreTransitiveClosure(plex.dm, 0, PETSC_TRUE,
                                              NULL, &closure))
    # Mark all remaining points as core
    CHKERR(DMLabelCreateIndex(lbl_owned, pStart, pEnd))
    for p in range(pStart, pEnd):
        CHKERR(DMLabelHasPoint(lbl_owned, p, &is_owned))
        CHKERR(DMLabelHasPoint(lbl_ghost, p, &is_ghost))
        if not is_ghost and not is_owned:
            CHKERR(DMLabelSetValue(lbl_core, p, 1))
```

So if this was copy-pasted, a big chunk went missing. In any case, what does the point SF for a DMSwarm actually look like? AFAICT, when I do DMCreateSwarm, a pointSF is never built.

Reply:
Mea culpa. I think I now remember removing this because I didn't think a vertex mesh could have cells in the halo or something.
If you modify this for a swarm you get:
```python
    # If any entity in closure(cell) is in the halo, then all those
    # entities in closure(cell) that are not in the halo are owned,
    # but not core.
    for c in range(cStart, cEnd):
        closure = c # by definition since just the vertex
        nclosure = 1
        is_owned = PETSC_FALSE
        for ci in range(nclosure):
            p = closure
            CHKERR(DMLabelHasPoint(lbl_ghost, p, &is_ghost))
            if is_ghost:
                is_owned = PETSC_TRUE
                break
        if is_owned:
            for ci in range(nclosure):
                p = closure
                CHKERR(DMLabelHasPoint(lbl_ghost, p, &is_ghost))
                if not is_ghost:
                    CHKERR(DMLabelSetValue(lbl_owned, p, 1))
```
which, since `nclosure=1`, simplifies down to
```python
    # If any entity in closure(cell) is in the halo, then all those
    # entities in closure(cell) that are not in the halo are owned,
    # but not core.
    for c in range(cStart, cEnd):
        closure = c # by definition since just the vertex
        is_owned = PETSC_FALSE
        p = closure
        CHKERR(DMLabelHasPoint(lbl_ghost, p, &is_ghost))
        is_owned = is_ghost
        if is_owned:
            CHKERR(DMLabelHasPoint(lbl_ghost, p, &is_ghost))
            if not is_ghost:
                CHKERR(DMLabelSetValue(lbl_owned, p, 1))
```
from which we can see `CHKERR(DMLabelSetValue(lbl_owned, p, 1))` will never execute.

> In any case, what does the point SF for a DMSwarm actually look like? AFAICT, when I do DMCreateSwarm, a pointSF is never built.

From memory a pointSF is set up for the swarm when you associate it with a cellDM, but I don't remember exactly. @ksagiyam and I had an extended slack chat about this that is now behind the message limit paywall - do you remember what we discussed?

Reply:
Can you just run the code and run `swarm.getPointSF().view()`?

I don't believe your statement about the cellDM because:

```c
PETSC_EXTERN PetscErrorCode DMSwarmSetCellDM(DM dm,DM dmcell)
{
  DM_Swarm *swarm = (DM_Swarm*)dm->data;

  PetscFunctionBegin;
  swarm->dmcell = dmcell;
  PetscFunctionReturn(0);
}
```

Reply:
I think back then we thought `celldm` was used whenever communication happens on `swarm`.
I'm not sure if `PointSF` is directly set in `swarm` somewhere. Can you try what @wence- is asking?

Reply:
With `test.py` containing
```python
from firedrake import *
m = UnitSquareMesh(1,1)
vertices = [(0., 0.), (0.1, 0.1), (0.9, 0.9), (1.0, 1.0)]
vm = VertexOnlyMesh(m, vertices)
vm._swarm.getPointSF().view()
vm.init()
vm._swarm.getPointSF().view()
V = FunctionSpace(vm, "DG", 0)
vm._swarm.getPointSF().view()
```
(just to check everything) I get
```
(firedrake) ➜  firedrake git:(0d-mesh) ✗ mpiexec -n 3 python test.py
PetscSF Object: 3 MPI processes
  type: basic
    sort=rank-order
  PetscSFSetGraph() has not been called yet
PetscSF Object: 3 MPI processes
  type: basic
    sort=rank-order
  PetscSFSetGraph() has not been called yet
PetscSF Object: 3 MPI processes
  type: basic
    sort=rank-order
  PetscSFSetGraph() has not been called yet
```
so @wence- you are correct, the SF graph never gets set (which I assume is the definition of what it means to build the SF).

Does it even make sense for mesh of vertices to have an SF? A PETSc SF is responsible for information exchange between MPI processes during assembly when we have a a halo region. But vertices have no spatial extent so do halos matter? `DMSwarmMigrate` does everything based off the cellDM - I also get rid of ghosts similarly.

Thoughts?

Reply:
> but we can come up with an interior facet integral of the parent mesh that involves some function on a vertex mesh, which would need halo data exchange

@ksagiyam can you explain a bit more what you mean by this?

Reply:
So an integral over an interior facet involves two cells (a and b) adjacent to the facet, but these cells might be owned by two different processes, say A and B. If process A assembles the interior facet integral, it has to talk to process B to get the data attached to cell b. If we had a function g defined on a vertex mesh, and g influences the interior facet integral, process A would also need to get values of g on cell b. So in this case holo data exchange on g has to happen.

Reply:
> Does it even make sense for mesh of vertices to have an SF? A PETSc SF is responsible for information exchange between MPI processes during assembly when we have a a halo region. But vertices have no spatial extent so do halos matter? DMSwarmMigrate does everything based off the cellDM - I also get rid of ghosts similarly.

Spatial extent is the wrong phrase, all of this stuff is topological connectivity. But I agree that there are no shared vertices in parallel distribution (every vertex is on exactly one process).

So step back from the specifics of the datastructure. What the SF is doing is encoding sharing information. It says "this local point (e.g. cell, face, ...) actually belongs to that remote processor".

By culling the points that live in ghost cells, you guarantee that there are never any shared points.

Integrating on vertex meshes (where there is no connectivity) doesn't need data from another process, since you ensured that vertices only live in cells we own.

Coupling the other way, from plex integrals onto vertices might need vertices from another process (as @ksagiyam describes above) since a face integral gets data from "both sides" one of which might be remote.

But I think you haven't thought about coupling like this yet?

So I would advocate culling the halo stuff from vertex meshes, everything is always core, and don't provide a halo object at all.

Reply:
Was I perhaps wrong to cull vertices from the ghost cells of the parent mesh? I did it because the counting didn't add up when I distributed the vertex only mesh but maybe if I labelled everything correctly with PyOP2 (to match the parent cell) it would just work when I, for example, assemble an integration over an entire distributed mesh?

Reply:
The issue was when using the DMSwarm's own methods like `getLocalSize`. @wence- what do you think?

Reply:
I'm not sure I understand what your model for parallelism and computation is, so I don't think I can answer that question. Can you explain what you think it is, and how distribution of the DMSwarm achieves it?

Reply:
> I'm not sure I understand what your model for parallelism and computation is, so I don't think I can answer that question. Can you explain what you think it is, and how distribution of the DMSwarm achieves it?

My aim: The `DMSwarm` is the `DM` for a `VertexOnlyMesh` in the same way that a `DMPlex` is for a `Mesh`. Parallelism and computation should be as similar as possible for a `Mesh` and `VertexOnlyMesh`.

Complicating factors come from the way in which a VertexOnlyMesh is created:
 1. I have a `Mesh` which is is distributed across ranks. 
    - There is an underlying `DMPlex` of topological entities which was distributed by calling `DMPlexDistribute`.
    - To allow correct halo data exchange during assembly on the mesh 
      - a PETSc `SF` is set up to make it clear which ranks own halo entities (I think) and
      - the appropriate entities are labelled as needed with PyOP2.
    - The `DMPlex` has various petsc4py methods/PETSc functions. Their behaviour wrt the halo is not completely clear to me.
       - I'm not sure if PETSc functions like `DMPlexGetHeightStratum` (which `mesh.num_cells` calls) include the halo when they count.  (@wence- maybe you can answer this)
       - I know that we use the PyOP2 labelling to count entities by using label-specific functions like `DMGetStratumSize`
       - I believe we only use the `SF` when we need to change the PyOP2 labelling but I'm not quite sure.
 2. I create a VertexOnlyMesh from the parent mesh and a collection of points. The points are not required to be the same across all ranks.
    - I create a Particle In Cell (PIC) `DMSwarm` with the DMPlex of the parent mesh it's `CellDM` by calling `DMSwarmSetCellDM`.
    - I let PETSc distribute the point list over the distributed `DMPlex` using the `DMSwarmSetPointCoordinates` function.
      - On each rank, points which are not anywhere on the entire distributed `DMPlex` are rejected. 
      - `DMPlex` cell entities which are in the halo get points put in them on each rank where they are present.
      - If I call `DMSwarmGetLocalSize` at this point on each rank and reduce the result I find that I double count all the points that were assigned to `DMPlex` halo region cells.
    - I then remove all points that are in the halo regions to allow `DMSwarm` functions like `DMSwarmGetLocalSize` to give me the expected result when I reduce the answer. **This may be wrong**.

Were I to _allow_ the double counting, could I make the `DMSwarm` based `VertexOnlyMesh` behave more like a `Mesh`? I worry I got sent on a wild goose chase when `DMSwarmGetLocalSize` gave me unexpected results and removing ghost particles was wrong. If I inherit the parent cell's PyOP2 labelling and ignore functions like `DMSwarmGetLocalSize` assembly should just work right? (I'd also need to set up the PETSc `SF` to make clear who owns what I think - the `dmplex.pyx` function `prune_sf` might prove a guide?)

Perhaps this is too complex and I should make setting all this up an aspiration for when I need it...

Reply:
To clarify by double counting I mean that if `DMSwarmSetPointCoordinates` puts a point into a cell in the ghost region, it does so on all MPI ranks. Therefore I will see the same particle more than once when I MPI reduce (for example). In hindsight I think this is correct and I need to just PyOP2 label them correctly.

Reply:
Sure, so this is a standard model. You have a global set of things that are distributed, sometimes they appear on different processes more than once, but only one process actually "owns" things (this is what the ghost cells are).

The SF is used _both for determining what is in the ghost region_, and _determining who the real owner of a ghost point is_. So that you can update your ghost value with correct data.

You need halos if you have coupling between topological entities in a mesh that crosses process boundaries. I think you do not have this, and therefore don't need a halo.

So I would:

1. Cull from ghost cells as you do now.
2. Don't pretend you have a pointSF (you don't), and make all your sets with `halo=None`
3. Label all topological entities (vertices) in your mesh as _core_. [make empty labels for owned and ghost]

Once you know how you want to do computation, then you can go back and put the halo stuff in. Notice that a correct SF is crucial for this, since it allows moving halo data around.

Reply:
I don't know if I'll implement it now, but given the a `VertexOnlyMesh` is by definition immersed in another mesh, isn't all the information I need to get halo exchange correct in the pointSF of the parent mesh's `DMPlex`? That's what PETSc uses in `DMSwarmSetCellDM` and I believe in `DMSwarmMigrate`.

Reply:
You're conflating two things, I think.

1. How to determine which cell (and process) a PIC point has moved to.
2. If I have a PIC point on process A in a ghost cell in the parent mesh, which PIC point that corresponds to on process B (which actually owns it).

The DMPlex's pointSF tells you, having located the process local cell you're in, how to do 1. But for point 2 you need more information (like is the data layout the same on both processes, how many PIC points are there in this cell, etc...). I think all the information is available, but it's not immediately trivial.

I continue to note that `DMSwarmSetCellDM` doesn't actually do anything other than remember the DMPlex object. `DMSwarmMigrate` moves PIC points that have moved out of the owned cells on the current process to their correct target new process, this is not the same as halo exchanging (although it is similar).

Reply:
> I don't know if I'll implement it now, but given the a `VertexOnlyMesh` is by definition immersed in another mesh, isn't all the information I need to get halo exchange correct in the pointSF of the parent mesh's `DMPlex`? That's what PETSc uses in `DMSwarmSetCellDM` and I believe in `DMSwarmMigrate`.

Addendum: this was meant to say 

> "That's what PETSc uses in _`DMSwarmSetPointCoordinates`_ and I believe in `DMSwarmMigrate`.

Reply:
Okay so tell me if I've got this right: the parent mesh `DMPlex` pointSF can tell us which parent mesh cell a given point is in, but if I've got multiple points per parent mesh cell, I don't know which of those points it is. So really I need to make a pointSF graph for the `DMSwarm` somehow. Or use the parent mesh `DMPlex` pointSF and add extra information by somehow tracking numbering within parent cells.

Reply:
Something like that, yes. But you only need to do this _if you need ghost cell PIC data_. Do you need ghost cell PIC data?

Reply:
I don't currently have plans to solve anything on a `VertexOnlyMesh` which I think is the only time this is relevant?

Reply:
> 1. Cull from ghost cells as you do now.
> 2. Don't pretend you have a pointSF (you don't), and make all your sets with `halo=None`
> 3. Label all topological entities (vertices) in your mesh as _core_. [make empty labels for owned and ghost]

So (1.) and (3.) are done, I'm not sure what modifications are required for (2.) If I do
```diff
diff --git a/firedrake/functionspacedata.py b/firedrake/functionspacedata.py
index 18e9fb48..e97de598 100644
--- a/firedrake/functionspacedata.py
+++ b/firedrake/functionspacedata.py
@@ -87,7 +87,11 @@ def get_node_set(mesh, key):
     nodes_per_entity, real_tensorproduct = key
     global_numbering = get_global_numbering(mesh, (nodes_per_entity, real_tensorproduct))
     node_classes = mesh.node_classes(nodes_per_entity, real_tensorproduct=real_tensorproduct)
-    halo = halo_mod.Halo(mesh._topology_dm, global_numbering)
+    # A VertexOnlyMeshTopology has an empty halo.
+    if type(mesh) is mesh_mod.VertexOnlyMeshTopology:
+        halo = None
+    else:
+        halo = halo_mod.Halo(mesh._topology_dm, global_numbering)
     node_set = op2.Set(node_classes, halo=halo, comm=mesh.comm)
     extruded = mesh.cell_set._extruded
```
then when running `py.test -x tests/vertexonly/test_vertex_only_fs.py` I get 
```
...
tests/vertexonly/test_vertex_only_fs.py:117: in test_functionspaces
    functionspace_tests(vm)
tests/vertexonly/test_vertex_only_fs.py:70: in functionspace_tests
    g.project(expr)
firedrake/adjoint/function.py:40: in wrapper
    output = project(self, b, *args, **kwargs)
firedrake/function.py:337: in project
    return projection.project(b, self, *args, **kwargs)
firedrake/adjoint/projection.py:29: in wrapper
    output = project(*args, **kwargs)
firedrake/projection.py:92: in project
    use_slate_for_inverse=use_slate_for_inverse).project()
firedrake/projection.py:170: in project
    self.apply_massinv(self.target, self.rhs)
firedrake/projection.py:156: in solve
    self.A.petscmat.mult(b_, x_)
../PyOP2/pyop2/utils.py:62: in __get__
    obj.__dict__[self.__name__] = result = self.fget(obj)
firedrake/projection.py:141: in A
    form_compiler_parameters=self.form_compiler_parameters)
firedrake/adjoint/assembly.py:17: in wrapper
    output = assemble(*args, **kwargs)
firedrake/assemble.py:100: in assemble
    for l in loops:
firedrake/assemble.py:704: in _assemble
    form_compiler_parameters=form_compiler_parameters)
firedrake/assemble.py:269: in get_matrix
    block_sparse=baij)
../PyOP2/pyop2/caching.py:157: in __new__
    obj = make_obj()
../PyOP2/pyop2/caching.py:138: in make_obj
    obj.__init__(*args, **kwargs)
../PyOP2/pyop2/base.py:2842: in __init__
    nnz, onnz = build_sparsity(self)
pyop2/sparsity.pyx:136: in pyop2.sparsity.build_sparsity
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = DataSet(Set((1, 1, 1), 'set_#x11eee4cd0'), (1,), 'None_nodes_dset'), name = 'scalar_lgmap'

    def __getattr__(self, name):
        """Returns a Set specific attribute."""
>       value = getattr(self.set, name)
E       AttributeError: 'Set' object has no attribute 'scalar_lgmap'

../PyOP2/pyop2/base.py:942: AttributeError
```

Reply:
Oh yeah, I forgot another place we need the halo. For insertion into matrices, we need a translation of _process-local_ numbering into _process-global_ numbering. Try this patch in PyOP2?

```
diff --git a/pyop2/petsc_base.py b/pyop2/petsc_base.py
index c70e78cf..473c9359 100644
--- a/pyop2/petsc_base.py
+++ b/pyop2/petsc_base.py
@@ -55,8 +55,10 @@ class DataSet(base.DataSet):
         indices for this :class:`DataSet`.
         """
         lgmap = PETSc.LGMap()
-        if self.comm.size == 1:
-            lgmap.create(indices=np.arange(self.size, dtype=IntType),
+        if self.comm.size == 1 or self.halo is None:
+            offset = self.comm.scan(self.size) - self.size
+            indices = np.arange(offset, offset + self.size, dtype=IntType)
+            lgmap.create(indices=indices,
                          bsize=self.cdim, comm=self.comm)
         else:
             lgmap.create(indices=self.halo.local_to_global_numbering,
```

Reply:
This says that for vertex meshes, the points are numbered globally first on process 0, then on process 1, etc...

Reply:
You could also implement a stub halo object for vertex only meshes in Firedrake (in halo.py) that does that, and mocks out the other functions with `pass`.

```
class VertexMeshHalo(op2.Halo):
    def __init__(self, set_):
        assert len(set(set_.sizes)) == 1
        self.sizes = set_.sizes
        self.comm = self.set_.comm
    @utils.cached_property
     def local_to_global_numbering(self):
        # code from patch above
     
     def noop_exchange(self, *args):
          pass
     global_to_local_begin = noop_exchange
     global_to_local_end = noop_exchange
     local_to_global_begin = noop_exchange
     local_to_global_end = noop_exchange
```

That's probably safer because now if your vertex mesh sets get ghosts, you'll hit that assertion and then remember to fix things up.

Reply:
> This says that for vertex meshes, the points are numbered globally first on process 0, then on process 1, etc...

Where are you inferring this from?

Reply:
> This says that for vertex meshes, the points are numbered globally first on process 0, then on process 1, etc...

Is there a way to guarantee this? `dmswarm.create_section` and `dmplex.create_section` don't obviously interact with mpi process rank. Perhaps I'm misunderstanding?

Reply:
> Is there a way to guarantee this? 

Unless the code I wrote has bugs, it's prescriptive.

> dmswarm.create_section and dmplex.create_section don't obviously interact with mpi process rank. Perhaps I'm misunderstanding?

I recommend reading https://arxiv.org/abs/1506.06194 and https://arxiv.org/abs/2004.08729, and maybe https://jedbrown.org/files/StarForest.pdf, the idea of petscsection for data layout is explained in section 7.2 of https://cse.buffalo.edu/~knepley/classes/caam519/CSBook.pdf

The basic parallel computation model is as follows:

1. A DM has some points that represent elements of the topology, when it is distributed each process gets some subset (not necessarily disjoint) of these points
2. `create_section` makes a _process-local_ datastructure that says how much data is associated with each point the local process can see.

In the case of a DMPlex, the subsets of points from step (1) are tied together in parallel by the "point SF" (`dm.getPointSF()`) which is a data structure that, for all shared points, says who owns them (and where they live [what they're point number is] on the owner process).

OK, so now we can bootstrap a consistent global numbering of all points in the parallel mesh. There is almost anything you can do here, you just need to uniquely assign a global number to each local number (without gaps). What is almost always done, and is done here, is to say that the numbering of owned points is contiguous and starts at `sum(nowned_points_of_processes_with_lower_rank_than_me)`. Having determined the numbering of all your owned points, you now need to figure out what the numbers of the unowned points are. This is done by broadcasting over the PetscSF that ties the points together.

For sections, exactly the same process happens, but now you number the dofs in the section contiguously, rather than the points in the mesh.

So now to your vertex only mesh. The point distribution amongst processes _is_ disjoint so you can get away with just producing this global contiguous numbering of mesh points. I guess if you wanted to go down the SF route (hence using the existing halo object) you just make an SF that advertises no shared points:

```python
sf = PETSc.SF().create(comm=dm.comm)
nroots = number_of_points_owned_on_my_process
# No leaves (remote points)
sf.setGraph(nroot, None, [])
```

Reply:
This is really useful, thanks. So I think I can bypass any special-case halo stuff by applying this patch:

```diff
diff --git a/firedrake/mesh.py b/firedrake/mesh.py
index cc1914c2..cf2ebe04 100644
--- a/firedrake/mesh.py
+++ b/firedrake/mesh.py
@@ -1963,6 +1963,13 @@ def _pic_swarm_in_plex(plex, coords, fields=[]):
     # point.
     dmswarm.remove_ghosts_pic(swarm, plex)
 
+    # Set the `SF` graph to advertises no shared points (since the halo
+    # is now empty) by setting the leaves to an empty list
+    sf = swarm.getPointSF()
+    nroots = swarm.getLocalSize()
+    sf.setGraph(nroots, None, [])
+    swarm.setPointSF(sf)
+
     return swarm
 
 

```
Furthermore, I think that `dmplex.mark_entity_classes` can be made generic for all `DM`s since the given SF will force all entities to be marked as `pyop2_core`. Do you agree?

Reply:
Argh except there's a bunch of `DMPlex` specific functions in `dmplex.mark_entity_classes`

Reply:
Okay I've generalised `mark_entity_classes` and `get_entity_classes` and moved them to `dmcommon.pyx`. Let me knwo what you think. I type check on the `DM` and do the appropriate thing in those two functions. I'm aware that more could be moved to `dmcommon.pyx` if I take this approach - what do you think?

Reply:
This is resolved - keeping listed as "unresolved" for future reference since there's some useful discussion here.

Reply:
See #1793